lib/os/shutdown.c

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/sys/shutdown_internal.h>
+#include <zephyr/toolchain.h>
+
+#if defined(CONFIG_MULTITHREADING) && !defined(CONFIG_SMP)
+#include <zephyr/kernel.h>
+#endif
+
+void sys_shutdown(void)
+{
+#if defined(CONFIG_MULTITHREADING) && !defined(CONFIG_SMP)
+	z_sys_shutdown_schedule();
+	k_sleep(K_FOREVER);
```

Comment:
```
pm_state_force(0u, &(struct pm_state_info){ PM_STATE_SOFT_OFF, 0, 0 });
k_sleep(K_FOREVER);
```

Basically ...

Reply:
yes, but just with `sys_shutdown()` call, so:

- No clunky inline structs (what are those 0?)
- No need to sleep outside (K_FOREVER? Some value and later a fallback loop "just in case it fails"...?)
- API gives you guarantees of the behavior, so you can write portable code
- No need to do hacks like avoiding soft-off in devicetree (just see comments in NXP files, or the recently removed em4 in silabs)
- No need to do hacks like `disable_ds_1` in Nordic system off sample

To me, this really makes a difference. If for some it doesn't, I guess I'm done with the PR.

Reply:
that is not what I meant, from user the API simplify things no doubt. Talking about the internal implementation that could be achieved doing that.

> K_FOREVER? Some value and later a fallback loop "just in case it fails"...?)
```
printk("ERROR: System off failed\n");
while (true) {
	/* spin to avoid fall-off behavior */
}
```
Could easily be replaced with `CODE_UNREACHABLE;` as it is done in ` sys_shutdown`. Just pointing that this pattern is similar.

> No need to do hacks like avoiding soft-off in devicetree (just see comments in NXP files, or the recently removed em4 in silabs)

That is a good point indeed.